metagamer/environments/qagent.py

```diff
@@ -1,144 +1,432 @@
-
-            num_steps[i] = steps
-            eps_vals[i] = epsilon
-
-            if i and i % test_window == 0:
-                # every 5%
-                upp = i
-                low = upp - test_window
-                print(
-                    f"{i}: eps:{epsilon:.2f},  max: {np.max(num_steps[low:upp])}"
-                    f" ave: {np.mean(num_steps[low:upp]):.2f}"
-                    f" std: {np.std(num_steps[low:upp]):.2f}"
+                    self.agent2.set_reward(p2_state, p2_action, p2_qscore)
+                else:
+                    self.agent1.set_reward(p1_state, p1_action, reward)
+                    self.agent2.set_reward(p2_state, p2_action, reward * -1)
+                    break
+
+            results[i] = reward
+
+        print(
+            f"games {num_epsiodes}"
+            f" crosses: {np.sum(results == 1):.2f}"
+            f" naughts: {np.sum(results == -1):.2f}"
+            f" draws: {np.sum(results == 0):.2f}"
+        )
+
+        return results
+
+    def run(self):
+        done = False
+        with self.agent1.exploit_mode(), self.agent2.exploit_mode():
+            self.env.reset()
+            while not done:
+                p1_state = self.agent1_env.get_observation()
+                logger.info(
+                    "p1_state: '%s', valid_actions: %s",
+                    p1_state,
+                    self.agent1_env.valid_actions,
+                )
+                p1_action = self.agent1.get_action(
+                    p1_state, self.agent1_env.valid_actions
+                )
+                logger.info("p1_action: %s", p1_action)
+
+                _, reward, done, _ = self.agent1_env.step(p1_action, 1)
+                logger.info("p1 reward: %s, done: %s", reward, done)
+
+                self.env.render("human")
+
+                if done:
+                    break
+
+                p2_state = self.agent2_env.get_observation()
+                p2_action = self.agent2.get_action(
+                    p2_state, self.agent2_env.valid_actions
                 )
+                _, reward, done, _ = self.agent2_env.step(p2_action, -1)
+
+                self.env.render("human")
+                if done:
+                    break
+            print(f"Result was : {reward}")
+
+        self.agent1.train = True
+        self.agent2.train = True
+
+
+def pure_self_play():
+    p1 = QTicTacTable(0.1)
+    p2 = QTicTacTable(0.1)
+    new_dome = TicTacToeRunner(p1, p2)
+    nr = 100
+    p2.train = True
+    p1.train = True
+    for i in range(nr):
+        p1.epsilon = (nr-1-i)/nr
+        p2.epsilon = (nr-1-i)/nr
+        print("epsilon", p1.epsilon, end=" ")
+        new_dome.train(100)
+
+    print("\nNo training")
+    p1.train = False
+    p2.train = False
+    new_dome.train(100)
+    new_dome.run()
+
+
+
+def main():
+
+
+
+    from pprint import pprint
+
+    p1 = QTicTacTable(0.1)
+    determined = PageLines()
+    battledome = TicTacToeRunner(p1, determined)
```

Comment:
Happy to leave the main as being a against a determined player?  I think it's fair enough, but we should write another couple more interesting ones.

Reply:
Yeah, this was more meant to be a few functions with different training regimens. The default main is simple and quick and also provides a nice starting point with examples of how to use the training regiment